### PR TITLE
Create ArgoCD clients on demand

### PR DESCRIPTION
**Pull request by raphink**
This PR aims to create ArgoCD clients on demand, in order to allow setting the
provider dynamically, based on other resource attributes.

The idea is to store the provider's `ResourceData` in the `meta` and use it to
initialize the clients when resources need them.

This PR is based on top of #62.

TODO:

- [x] pass the provider's `ResourceData` as pointer to resources
- [x]  call `initClients()` in all resource types
- [x]  cache clients to avoid duplication

### CODE REVIEW

**Comment by raphink** [comment]
@oboukili what is the status for #62?

**Comment by oboukili** [comment]
Unfortunately, there is no answer from the PR author of #62

**Comment by ClaireBellivier** [comment]
If you can wait for next Tuesday, I can have a look at the remaining tasks. Otherwise, feel free to complete the pull request yourself.

**Comment by raphink** [comment]
Great. I'm also happy to help if necessary 😁

**Comment by raphink** [comment]
Stale status: waiting on #62

**Comment by oboukili** [comment]
Hi @raphink , #62 has just been merged, therefore if you could update your branch from the master branch, we can resume the work on this PR.
I will hold off releasing any new version until we merge your PR as this one in addition to #62 warrants a major version bump of the provider.

**Comment by raphink** [comment]
@oboukili Thanks. I'm going to try and rebase. I'll let you know when it's ready.

**Comment by mcanevet** [comment]
@oboukili I tested this PR. It works fine.

**Review by mcanevet — APPROVED** [review]

**Comment by raphink** [comment]
@oboukili this should be ready

File: GNUmakefile
Comment (oboukili) [inline GNUmakefile:46]: the script gofmtcheck.sh does not seem to exist, can you add it?

File: GNUmakefile
Comment (oboukili) [inline GNUmakefile:65]: These helper scripts make directives (testacc*) are useful to spin up a test cluster, at least I use them a lot :) can you leave them within the Makefile?

**Review by oboukili — CHANGES_REQUESTED** [review]

**Comment by raphink** [comment]
@oboukili is it ok like this?

**Comment by oboukili** [comment]
Yes, we just need to wait for the CI tests to pass, I think I noticed a test case error earlier. Let's see.

**Comment by oboukili** [comment]
Apparently there's a race condition 
```
/home/runner/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.7.1/helper/schema/resource_data.go:84 +0x5f fp=0xc00395b378 sp=0xc00395b2c8 pc=0x2dad55f
github.com/oboukili/terraform-provider-argocd/argocd.initApiClient(0xc000cecf80, 0x0, 0x0, 0x0, 0x0)
	/home/runner/work/terraform-provider-argocd/terraform-provider-argocd/argocd/provider.go:148 +0x2c6 fp=0xc00395b520 sp=0xc00395b378 pc=0x32081c6
github.com/oboukili/terraform-provider-argocd/argocd.(*ServerInterface).initClients(0xc000b599f0, 0x0, 0x0)
	/home/runner/work/terraform-provider-argocd/terraform-provider-argocd/argocd/features.go:50 +0x8a5 fp=0xc00395b608 sp=0xc00395b520 pc=0x3206de5
github.com/oboukili/terraform-provider-argocd/argocd.resourceArgoCDProjectTokenDelete(0x3d6f130, 0xc001c6e240, 0xc000ceda00, 0x3549560, 0xc000b599f0, 0xc003032e50, 0xc00395b7b0, 0x10fdd78)
	/home/runner/work/terraform-provider-argocd/terraform-provider-argocd/argocd/resource_argocd_project_token.go:420 +0x5b fp=0xc00395b750 sp=0xc00395b608 pc=0x321567b
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).delete(0xc001bc3180, 0x3d6f0c0, 0xc0025fdf80, 0xc000ceda00, 0x3549560, 0xc000b599f0, 0x0, 0x0, 0x0)
```

I never had the opportunity to debug this so I can't really offer serious guidance, but I remember `go test -race` might come in handy.

**Comment by raphink** [comment]
I'm guessing this might have to do with multiple resources trying to init the clients at the same time. I could probably add a flag on the clients to mark them as initialized and avoid this behavior.

**Comment by oboukili** [comment]
We should use some kind of locking mechanism such as a sync.RWMutex or try to use a sync.syncMap.
The former will be a bit tedious to add, the latter can be dangerous (never used it) as apparently some guarantees that Go provides get thrown out the window like type safety.

If you use the RWMutex, I think it should be located out of the clients' objects.

**Comment by raphink** [comment]
NOTE: @oboukili I used a `sync.Mutex` instead of a `sync.RWMutex` because we don't want to take the risk that a read-only client would exit the `initClients()` function before the first (write) client has initialized everything.

**Comment by oboukili** [comment]
@raphink the tests timeout quickly now, I believe it's because your lock is an exclusive one, not a RWMutex one, therefore even for a read a lock is positioned. Please try and use a RWMutex lock, thanks.

File: argocd/features.go
Comment (oboukili) [inline argocd/features.go:46]: This is not enough, the locking mechanism has to be updated as well in the function with RLock/Lock and RUnlock/Unlock

**Review by oboukili — CHANGES_REQUESTED** [review]

**Comment by raphink** [comment]
@oboukili I'm probably misunderstanding something I guess. If I use an RLock, then some resources will be able to read the `ServerInterface` before it is done initializing. I don't see a way to ensure all resources have fully initialized clients without an exclusive lock. How would I do that?

**Comment by oboukili** [comment]
Extremely naively (sorry I don't have so much time to spare at the moment) I would say:

Modify the initClients() function to:
```go
func (p *ServerInterface) initClients() error {
	if p.Initialized {
		return nil
	}
	p.InitLock.Lock()

	d := p.ProviderData
...
```

and position a RLock very early on on every resource CRUD function:
```go
func resourceArgoCDApplicationCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
	objectMeta, spec, diags := expandApplication(d)
	if diags != nil {
		return diags
	}
	server := meta.(*ServerInterface)
	server.InitLock.RLock()
	defer server.InitLock.RUnlock()
       ...
```

Good luck and thanks!

**Comment by oboukili** [comment]
A syncMap would be most interesting though! If applicable.

**Comment by raphink** [comment]
The problem is not really with a map though (or is it?) but with the various fields of the `ServerInterface` struct. Are these managed as a map internally?

**Comment by raphink** [comment]
@oboukili I managed to get it to pass (at least locally) with a simple mutex by just using `defer`, which I hadn't done in the first try. Could you just launch the pipeline to see if it's green like this please?

**Comment by raphink** [comment]
I see the acceptance tests are now passing. @oboukili is that good for you?

**Comment by oboukili** [comment]
Oh great! Awesome work @raphink ! Let me merge that and make a major release ;)

**Pull request merged**
